pkg/resourceinterpreter/customized/webhook/serviceresolvers.go

```diff
@@ -0,0 +1,80 @@
+package webhook
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	webhookutil "k8s.io/apiserver/pkg/util/webhook"
+	corev1 "k8s.io/client-go/listers/core/v1"
+)
+
+// ServiceResolver knows how to convert a service reference into an actual location.
+type ServiceResolver interface {
+	ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
+}
+
+// NewServiceResolver returns a ServiceResolver that parses service first,
+// if service not exist, constructs a service URL from a given namespace and name.
+func NewServiceResolver(services corev1.ServiceLister) ServiceResolver {
+	return &serviceResolver{
+		services:        services,
+		defaultResolver: webhookutil.NewDefaultServiceResolver(),
+	}
+}
+
+type serviceResolver struct {
+	services        corev1.ServiceLister
+	defaultResolver ServiceResolver
+}
+
+func (r *serviceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
+	res, err := resolveCluster(r.services, namespace, name, port)
+	if err != nil && apierrors.IsNotFound(err) {
+		return r.defaultResolver.ResolveEndpoint(namespace, name, port)
+	}
+	return res, err
+}
```

Comment:
We can get the service first and 
- if service is not found, then construct service URL from namespace and name.
- if service exists, then construct the service URL from the external URL or cluster IP.

Reply:
updated.